File: MvvmSample/app/src/main/java/com/abben/mvvmsample/repository/NetworkBoundResource.java
package com.abben.mvvmsample.repository;


import android.arch.lifecycle.LiveData;
import android.arch.lifecycle.MediatorLiveData;
import android.arch.lifecycle.Observer;
import android.support.annotation.MainThread;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.annotation.WorkerThread;

import com.abben.mvvmsample.network.ApiResponse;
import com.abben.mvvmsample.util.Objects;
import com.abben.mvvmsample.vo.Resource;

import io.reactivex.Observable;
import io.reactivex.ObservableSource;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.functions.Consumer;
import io.reactivex.functions.Function;
import io.reactivex.schedulers.Schedulers;

/**
 * 从SQLite数据库和网络获得资源数据
 * Created by abben on 2017/10/24.
 * <p>
 * 更多信息请参考<a href="https://developer.android.com/arch">Architecture
 * Guide</a>.
 *
 * @param <ResultType>  数据库以及APP所使用的资源数据的类型
 * @param <RequestType> 远程返回的资源数据的类型
 */
public abstract class NetworkBoundResource<ResultType, RequestType> {


    private final MediatorLiveData<Resource<ResultType>> result = new MediatorLiveData<>();

    @MainThread
    public NetworkBoundResource() {
        result.setValue((Resource<ResultType>) Resource.loading(null));
        final LiveData<ResultType> dbSource = loadFromDb();
        //Note: As of version 1.0, Room uses the list of tables accessed in the query to decide whether to update instances of LiveData.
        result.addSource(dbSource, new Observer<ResultType>() {
            @Override
            public void onChanged(@Nullable ResultType resultType) {
                result.removeSource(dbSource);
                if (shouldFetch(resultType)) {
                    fetchFromNetwork(dbSource);
                } else {
                    result.addSource(dbSource, new Observer<ResultType>() {
                        @Override
                        public void onChanged(@Nullable ResultType resultType) {
                            setValue(Resource.success(resultType));
                        }
                    });
                }
            }
        });
    }

    @MainThread
    private void setValue(Resource<ResultType> newValue) {
        if (!Objects.equals(result.getValue(), newValue)) {
            result.setValue(newValue);
        }
    }

    private void fetchFromNetwork(final LiveData<ResultType> dbSource) {
        final LiveData<ApiResponse<RequestType>> apiResponse = createCall();
        //我们将dbsource作为一种新数据来源，它将很快地发出它的最后一个数据
        result.addSource(dbSource, new Observer<ResultType>() {
            @Override
            public void onChanged(@Nullable ResultType resultType) {
                setValue(Resource.loading(resultType));
            }
        });
        result.addSource(apiResponse, new Observer<ApiResponse<RequestType>>() {
            @Override
            public void onChanged(@Nullable final ApiResponse<RequestType> requestTypeApiResponse) {
                result.removeSource(apiResponse);
                result.removeSource(dbSource);
                //检查数据加载状态
                if (requestTypeApiResponse.isSuccessful()) {
                    //保存结果然后重新初始化
                    Observable.just(processResponse(requestTypeApiResponse))
                            .flatMap(new Function<RequestType, ObservableSource<Boolean>>() {
                                @Override
                                public ObservableSource<Boolean> apply(@io.reactivex.annotations.NonNull RequestType requestType) throws Exception {
                                    saveCallResult(requestType);
                                    return Observable.just(true);
                                }
                            })
                            .subscribeOn(Schedulers.io())
                            .observeOn(AndroidSchedulers.mainThread())
                            .subscribe(new Consumer<Boolean>() {
                                @Override
                                public void accept(Boolean aBoolean) throws Exception {
                                    //我们特别要求一个新的实时数据，
                                    //否则，我们将立即获得上次缓存的值，
                                    //可能不会从网络接收到最新结果。
                                    result.addSource(loadFromDb(), new Observer<ResultType>() {
                                        @Override
                                        public void onChanged(@Nullable ResultType resultType) {
                                            setValue(Resource.success(resultType));
                                        }
                                    });
                                }
                            });

                } else {
                    onFetchFailed();
                    result.addSource(dbSource, new Observer<ResultType>() {
                        @Override
                        public void onChanged(@Nullable ResultType resultType) {
                            setValue(Resource.error(requestTypeApiResponse.errorMessage, resultType));
                        }
                    });
                }
            }
        });
    }


    @WorkerThread
    protected RequestType processResponse(ApiResponse<RequestType> response) {
        return response.body;
    }


    /**
     * 调用它将API响应的结果保存到数据库中。
     *
     * @param item
     */
    @WorkerThread
    protected abstract void saveCallResult(@NonNull RequestType item);

    /**
     * 调用它来决定是否应该从网络获取数据。
     *
     * @param data
     * @return
     */
    @MainThread
    protected abstract boolean shouldFetch(@Nullable ResultType data);

    /**
     * 调用它从数据库获取缓存的数据。
     *
     * @return
     */
    @NonNull
    @MainThread
    protected abstract LiveData<ResultType> loadFromDb();

    /**
     * 调用它来创建API调用。
     *
     * @return
     */
    @NonNull
    @MainThread
    protected abstract LiveData<ApiResponse<RequestType>> createCall();

    /**
     * 获取失败时调用。子类可能需要重置组件。
     * 类似rate limiter(速率限制器)。
     */
    protected abstract void onFetchFailed();

    protected abstract void onFetchSuccess();

    public LiveData<Resource<ResultType>> asLiveData() {
        onFetchSuccess();
        return result;
    }
}

File: MvvmSampleJetpack/app/src/main/java/com/abben/mvvmsamplejetpack/common/NetworkBoundResource.kt
package com.abben.mvvmsamplejetpack.common

import androidx.annotation.MainThread
import androidx.annotation.WorkerThread
import androidx.lifecycle.LiveData
import androidx.lifecycle.MediatorLiveData
import com.abben.mvvmsamplejetpack.network.ApiEmptyResponse
import com.abben.mvvmsamplejetpack.network.ApiErrorResponse
import com.abben.mvvmsamplejetpack.network.ApiResponse
import com.abben.mvvmsamplejetpack.network.ApiSuccessResponse
import com.abben.mvvmsamplejetpack.vo.Resource
import io.reactivex.Observable
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.schedulers.Schedulers

/**
 *  @author abben
 *  on 2019/1/17
 */
abstract class NetworkBoundResource<ResultType, RequestType>
@MainThread constructor() {

    private val result: MediatorLiveData<Resource<ResultType>> = MediatorLiveData()

    init {
        result.value = Resource.loading(null)
        val dbSource: LiveData<ResultType> = loadFromDb()
        result.addSource(dbSource) {
            result.removeSource(dbSource)
            if (shouldFetch(it)) {
                fetchFromNetwork(dbSource)
            } else {
                result.addSource(dbSource) { newData ->
                    setValue(Resource.success(newData))
                }
            }
        }
    }

    @MainThread
    private fun setValue(newValue: Resource<ResultType>) {
        if (result.value != newValue) {
            result.value = newValue
        }
    }

    private fun fetchFromNetwork(dbSource: LiveData<ResultType>) {
        val apiResponse = createCall()
        //我们将dbsource作为一种新数据来源，它将很快地发出它的最后一个数据
        result.addSource(dbSource) {
            setValue(Resource.loading(it))
        }
        result.addSource(apiResponse) { response ->
            result.removeSource(apiResponse)
            result.removeSource(dbSource)
            //检查数据加载状态
            when (response) {
                is ApiErrorResponse -> {
                    onFetchFailed()
                    result.addSource(dbSource) {
                        setValue(Resource.error(response.errorMessage, it))
                    }
                }
                is ApiEmptyResponse -> {
                    result.addSource(loadFromDb()) {
                        setValue(Resource.success(it))
                    }
                }
                is ApiSuccessResponse -> {
                    Observable.just(processResponse(response))
                        .flatMap {
                            saveCallResult(it)
                            Observable.just(true)
                        }
                        .subscribeOn(Schedulers.io())
                        .observeOn(AndroidSchedulers.mainThread())
                        .subscribe {
                            //我们特别要求一个新的实时数据，
                            //否则，我们将立即获得上次缓存的值，
                            //可能不会从网络接收到最新结果。
                            result.addSource(loadFromDb()) { newData ->
                                setValue(Resource.success(newData))
                            }
                        }
                }
            }
        }
    }

    @WorkerThread
    protected fun processResponse(response: ApiSuccessResponse<RequestType>): RequestType {
        return response.body
    }

    /**
     * 调用它将API响应的结果保存到数据库中。
     *
     * @param item 从网络请求获得的数据
     */
    @WorkerThread
    protected abstract fun saveCallResult(item: RequestType)

    /**
     * 调用它来决定是否应该从网络获取数据。
     *
     * @param data 旧数据
     * @return 真就从网络获取数据，假就从数据库获取数据
     */
    @MainThread
    protected abstract fun shouldFetch(data: ResultType): Boolean

    /**
     * 调用它从数据库获取缓存的数据。
     *
     * @return 从数据库获得的数据
     */
    @MainThread
    protected abstract fun loadFromDb(): LiveData<ResultType>

    /**
     * 调用它来创建API调用。
     *
     * @return 从网络接口获得的数据
     */
    @MainThread
    protected abstract fun createCall(): LiveData<ApiResponse<RequestType>>

    /**
     * 获取失败时调用。子类可能需要重置组件。
     * 类似rate limiter(速率限制器)。
     */
    protected abstract fun onFetchFailed()

    protected abstract fun onFetchSuccess()

    fun asLiveData(): LiveData<Resource<ResultType>> {
        onFetchSuccess()
        return result
    }
}